### ISSUE TITLE
Definition type bug in branded types since 5.1

### ISSUE THREAD

**Issue by GabeAtWork** [microsoft/TypeScript#54655]
# Bug Report

Code that used to compile fine in 5.0.0 now produces invalid compiled code with type names suffixed in `_1` which are not defined (see code below).

### 🔎 Search Terms

Branded type, suffix, emit

### 🕗 Version & Regression Information

- This changed between versions 5.0.0 and 5.1.3

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/KYDwDg9gTgLgBDAnmYcBCUCGA7AJgHgCg51MBnYAGmPSzwEtsBzakgJWAqgDdhcA5TAFtUoGMDxk4ZGFEZM4AXjgAiAPpqkKDSsIA+JaQpwAZHADecANoBpOIzgcuvAcOABdAFy0cueXABfUws4DQB3ehhsTjINbzRyVACAbkJCUEhYBGRUNAgAVzw+AElscSZgKCISABkIMMq8wtw4MQlcKWx8oQAjStY4AFUwFCgmvFaQcUk4Lt7K-UMMX3w5vqhKVXHcErLgCqgVPVT08Gh4AGMIbBkECG3d8sqlGnw6hrGCibaZtf6hkaNL4tH4dWbddZ6AAUPWBZG85hoJAANvUgc1vO90XhUiQSPlAZ8MQDRttcYEAJRKPQ0KFI2beP5QGgU+LAx77SpvNFEvCbYak4EGRQ0kgAejFcGKQjAyOAIjKmBg9GucFwEE42AA5PAhErxFA4AALSrAelQ8xBchwHCIClAA)

### 💻 Code

```ts
// Source
export type Brand<
  Base,
  Branding,
  ReservedName extends string = "__type__"
> = Base & { [K in ReservedName]: Branding } & { __witness__: Base };

export type BoundedInteger<
  LowerBound extends number,
  UpperBound extends number
> = Brand<number, "BoundedInteger">;

export const toBoundedInteger =
  <LowerBound extends number, UpperBound extends number>(bounds: {
    lowerBound: LowerBound;
    upperBound: UpperBound;
  }) =>
  (
    n: number
  ): BoundedInteger<LowerBound, UpperBound> =>
  // Implementation doesn't matter here
    ({} as any)


```

### 🙁 Actual behavior

```ts
// v5.1.3

export type Brand<Base, Branding, ReservedName extends string = "__type__"> = Base & {
    [K in ReservedName]: Branding;
} & {
    __witness__: Base;
};
export type BoundedInteger<LowerBound extends number, UpperBound extends number> = Brand<number, "BoundedInteger">;
export declare const toBoundedInteger: <LowerBound extends number, UpperBound extends number>(bounds: {
    lowerBound: LowerBound;
    upperBound: UpperBound;
// ❌ Cannot find name 'LowerBound_1'. Did you mean 'LowerBound'?
}) => (n: number) => BoundedInteger<LowerBound_1, UpperBound_1>;
```

### 🙂 Expected behavior

```ts
// v5.0.4

export type Brand<Base, Branding, ReservedName extends string = "__type__"> = Base & {
    [K in ReservedName]: Branding;
} & {
    __witness__: Base;
};
export type BoundedInteger<LowerBound extends number, UpperBound extends number> = Brand<number, "BoundedInteger">;
export declare const toBoundedInteger: <LowerBound extends number, UpperBound extends number>(bounds: {
    lowerBound: LowerBound;
    upperBound: UpperBound;
}) => (n: number) => BoundedInteger<LowerBound, UpperBound>;
```

**Comment by fatcerberus** [comment]
I'm confused as to why there are type annotations in "compiled" code, in *any* TS version.  Even your Playground shows that the compiled JS code is:

```ts
export const toBoundedInteger = (bounds) => (n) => 
// Implementation doesn't matter here
({});
```

**Comment by GabeAtWork** [comment]
@fatcerberus in the playground, you can see the definitions in the `D.TS` tab:
![image](https://github.com/microsoft/TypeScript/assets/29233647/49eff448-6732-4682-9018-ea6bb0e72915)

Those definitions are exported in the emitted code so that TS code that imports from this compiled code (e.g. when this is a library) can have the proper types.

So for instance, I could emit the `.js` code along with the `.d.ts` code of this library which would enable me to import the type `BoundedInteger` from another project's code as well as having `toBoundedInteger` have proper types instead of `any`.

**Comment by fatcerberus** [comment]
Ah, you meant the `.d.ts`.  When you say "invalid compiled code" that makes it sound like you mean the output JS.  If you had said "invalid type definitions" or "invalid declarations"  that would have been clearer.

**Comment by GabeAtWork** [comment]
@fatcerberus good point, I've updated the issue description to hopefully clarify that, thanks for the feedback 🙏

**Comment by GabeAtWork** [comment]
Any news on this topic? The fix seems has been available since June, anything I can help with to help get it merged?

**Comment by jakebailey** [comment]
Using the original playground link, it looks like this bug is fixed in nightly.

I bisected and figured out that I fixed it in #55820... 😅 

I'll send a PR to add a test.
